controlplane/kubeadm/internal/workload_cluster_etcd.go

```diff
@@ -289,3 +289,73 @@ func (w *Workload) ForwardEtcdLeadership(ctx context.Context, machine *clusterv1
 	}
 	return nil
 }
+
+// EtcdStatus returns the current status of the etcd cluster
+func (w *Workload) EtcdStatus(ctx context.Context) (EtcdStatus, error) {
+	nodes, err := w.getControlPlaneNodes(ctx)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to list control plane nodes")
+	}
+
+	etcdClient, err := w.etcdClientGenerator.forNodes(ctx, nodes.Items)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to create etcd client")
+	}
+
+	members, err := etcdClient.Members(ctx)
+	if err != nil {
+		return EtcdStatus{}, errors.Wrap(err, "failed to list etcd members using etcd client")
+	}
+
+	statuses := map[string]EtcdMemberStatus{}
+	for _, member := range members {
+		status := EtcdMemberStatus{Member: member, Responsive: true}
+
+		_, err := etcdClient.EtcdClient.Status(ctx, staticPodName("etcd", member.Name))
+		if err != nil {
+			status.Responsive = false
+			w.Log.Error(err, "error fetching etcd status for member", "member", member.Name)
+		}
+		statuses[member.Name] = status
+	}
+
+	return EtcdStatus{
+		Members: statuses,
+	}, nil
+}
+
+// EtcdStatus is a snapshot of the etcd cluster's status
+type EtcdStatus struct {
+	Members map[string]EtcdMemberStatus
+}
+
+// EtcdMemberStatus is a snapshot of a etcd member's status
+type EtcdMemberStatus struct {
+	*etcd.Member
+	Responsive bool
+}
+
+// FailureTolerance is the amount of members the etcd cluster can afford to
+// lose without losing quorum.
+// Ref: https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md#what-is-failure-tolerance
+func (e *EtcdStatus) FailureTolerance(machine *clusterv1.Machine) int {
```

Comment:
If I look at the test below the method seems to work properly, but TBH I have some trouble in understanding the logic as it is written, because I'm not sure it is correct to exclude not responsive members from the size of the cluster.
 
In my mind, I would expect that:
- as a first step we should determine the size of the target cluster as the current number of members -1 (-1 should be applied only if the machine to be deleted has a corresponding member)
- given the size of the target cluster, I can compute the default tolerance
- non-responsive members should reduce the default tolerance (except for the machine to be deleted, which was already excluded from the size computation)

Does this sound reasonable to you?

Reply:
I don't think so, you're saying it's not correct to exclude non-responsive members from the size of the cluster, but then you're saying that they should reduce the default tolerance. It can't be both. 

The failure tolerance is `n - Floor(n / 2 + 1)`, where n is size of the cluster. Go automatically does the `Floor` when you convert to an int. In order to do what you're asking in your third bullet, you reduce `n`. 

I think you're proposing is that we do instead `(n - u - Floor(n / 2 +1))`, where `u` is the number of unresponsive members and we treat `n` as the total members, ignoring whether responsive or not. The problem with that is that you can then end up with negative numbers which aren't correct.

It works if you're just thinking of 1 unresponsive, `3 - 1 - 2 = 0`, which is correct. But if you had all 3 unresponsive you'd get `3 - 3 - 2 = -2`.

Reply:
Thinking about it more last night and realized that my formulation falls over with larger cluster sizes. You get stuff like a 9 member cluster with 2 unhealthy members yielding a failure tolerance of 3 instead of 2. The negative stuff is only relevant if you want to be able to say what the failure tolerance is of a cluster that's already exceeded it, and it's still sort of correct to say it's negative because it's been exceeded.

Reply:
What I'm proposing is to:
1. agree on how we determine n (the size of the target cluster)

Reply:
> What I'm proposing is to:
> 
> 1. agree on how we determine n (the size of the target cluster)
>    currently we are counting responsive members, and IMO this is not right.
>    so I propose to count all the members and reducing this value by one if the machine we are considering has one of those members

+1, we can take the information directly from etcd for the current cluster size, based on the call to `etcdClient.Members(ctx)`. If the Machine to be remediated is one of those members, we reduce by 1.

1a. If the Machine is not a member of etcd, it is safe to remediate

> 2. to determine the default/theoretical tolerance t  = n - Floor(n / 2 + 1)
> 3. to compute the available tolerance by reducing this number by the number of unhealthy machines (except the one I'm deleting)

To clarify, unhealthy here should be unresponsive to the etcdClient.EtcdClient.Status(), but should we also take into account any cluster wide alarms here as well (https://etcd.io/docs/v3.2.17/op-guide/maintenance/#space-quota)? 

> 
> If the available tolerance is >0, it is safe to remediate the node

Reply:
Maybe you didn't see but I updated it when I posted my comment earlier to use total members as `n`. I think we're in agreement if you re-review the method

Reply:
Thanks, I did indeed miss that update. I do think it might be good to take into account any cluster-wide alarms, especially in the case that we are bumping against the space quota.

Reply:
Hmm yeah I ran through a test where I filled up the etcd using [https://etcd.io/docs/v3.2.17/op-guide/maintenance/](this space quota tutorial) but it never triggered any alarms and so I thought it'd just be dead code. I don't know that we enable any alarms.